go/ql/lib/semmle/go/security/HardcodedCredentials.qll

```diff
@@ -0,0 +1,158 @@
+
+  private module Config implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+    predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    predicate isBarrier(DataFlow::Node node) { node instanceof Sanitizer }
+  }
+
+  /** Tracks taint flow for reasoning about hardcoded credentials. */
+  module Flow = TaintTracking::Global<Config>;
+
+  /** A hardcoded string literal as a source for hardcoded credentials. */
+  private class HardcodedStringSource extends Source {
+    HardcodedStringSource() { this.asExpr() instanceof StringLit }
+  }
+
+  /** A use of a credential. */
+  private class CredentialsSink extends Sink {
+    CredentialsSink() { exists(string s | s.matches("credentials-%") | sinkNode(this, s)) }
+  }
+
+  /**
+   * Holds if the guard `g` in its branch `branch` validates the expression `e`
+   * by comparing it to a disallowed literal.
+   */
+  private predicate constantValueCheck(DataFlow::Node g, Expr e, boolean branch) {
+    exists(Literal lit, DataFlow::EqualityTestNode eq | eq = g |
+      eq.getAnOperand().asExpr() = e and
+      eq.getAnOperand().asExpr() = lit and
+      e != lit and
+      branch = eq.getPolarity().booleanNot()
+    )
+  }
+
+  /**
+   * A value validated by comparing it to a disallowed constant value.
+   * For example, in the context `if key != "invalid_key" { ... }`,
+   * if `"invalid_key"` is indeed the only dangerous key then guarded uses of `key` are likely
+   * to be safe.
+   */
+  private class CompareExprSanitizer extends Sanitizer {
+    CompareExprSanitizer() {
+      this = DataFlow::BarrierGuard<constantValueCheck/3>::getABarrierNode()
+    }
+  }
+
+  /**
+   * A value returned with an error.
+   *
+   * Typically this means contexts like `return "", errors.New("Oh no")`,
+   * where we can be reasonably confident downstream users will not mistake
+   * that empty string for a usable key.
+   */
+  private class ReturnedAlongsideErrorSanitizer extends Sanitizer {
+    ReturnedAlongsideErrorSanitizer() {
+      exists(ReturnStmt r, DataFlow::CallNode c |
+        c.getTarget().hasQualifiedName("errors", "New") and
+        r.getNumChild() > 1 and
+        r.getAChild() = c.getAResult().getASuccessor*().asExpr() and
+        r.getAChild() = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * A value returned alongside an error-value that is known
+   * to be non-nil by virtue of a guarding check.
+   *
+   * For example, `if err != nil { return "", err }` is unlikely to be
+   * contributing a dangerous hardcoded key.
+   */
+  private class ReturnedAlongsideErrorSanitizerGuard extends Sanitizer {
+    ReturnedAlongsideErrorSanitizerGuard() {
+      exists(ControlFlow::ConditionGuardNode guard, SsaWithFields errorVar, ReturnStmt r |
+        guard.ensuresNeq(errorVar.getAUse(), Builtin::nil().getARead()) and
+        guard.dominates(this.getBasicBlock()) and
+        r.getExpr(1) = errorVar.getAUse().asExpr() and
```

Comment:
Nitpick: `return value, err` is common but not universal -- consider generalising this and the above to "the one return value with error type vs. the other non-error-typed returns"

Reply:
Good point. Turns out we have `DataFlow::returnedWithError`, so this can be greatly simplified: https://github.com/github/codeql/pull/15527/commits/d17defb00eae489ac3854f2fb9fab8a089906094.